include/power_grid_model/math_solver/iterative_current_pf_solver.hpp

```diff
@@ -0,0 +1,182 @@
+        IdxVector const& load_gen_bus_indptr = *this->load_gen_bus_indptr_;
+        IdxVector const& source_bus_indptr = *this->source_bus_indptr_;
+        std::vector<LoadGenType> const& load_gen_type = *this->load_gen_type_;
+
+        // set rhs to zero for iteration start
+        std::fill(rhs_.begin(), rhs_.end(), ComplexValue<sym>{0.0});
+
+        // rhs = I_inj + L'U
+        // loop buses: i
+        for (Idx bus_number = 0; bus_number != this->n_bus_; ++bus_number) {
+            // loop loads/generation: j
+            for (Idx load_number = load_gen_bus_indptr[bus_number]; load_number != load_gen_bus_indptr[bus_number + 1];
+                 ++load_number) {
+                // load type
+                LoadGenType const type = load_gen_type[load_number];
+                switch (type) {
+                    case LoadGenType::const_pq:
+                        // I_inj_i = conj(S_inj_j/U_i) for constant PQ type
+                        rhs_[bus_number] += conj(input.s_injection[load_number] / u[bus_number]);
+                        break;
+                    case LoadGenType::const_y:
+                        // I_inj_i = conj((S_inj_j * abs(U_i)^2) / U_i) = conj((S_inj_j) * U_i for const impedance type
+                        rhs_[bus_number] += conj(input.s_injection[load_number]) * u[bus_number];
+                        break;
+                    case LoadGenType::const_i:
+                        // I_inj_i = conj(S_inj_j*abs(U_i)/U_i) for const current type
+                        rhs_[bus_number] += conj(input.s_injection[load_number] * cabs(u[bus_number]) / u[bus_number]);
+                        break;
+                    default:
+                        throw MissingCaseForEnumError("Injection current calculation", type);
+                }
+            }
+            // loop sources: j
+            for (Idx source_number = source_bus_indptr[bus_number]; source_number != source_bus_indptr[bus_number + 1];
+                 ++source_number) {
+                // -L'U = Y_source_j * U_ref_j
+                rhs_[bus_number] += dot(y_bus.math_model_param().source_param[source_number],
+                                        ComplexValue<sym>{input.source[source_number]});
+            }
+        }
+    }
+
+    void solve_matrix() {
+        bsr_solver_.solve(mat_data_.data(), rhs_.data(), updated_u_.data(), true);
+    }
+
+    double iterate_unknown(ComplexValueVector<sym>& u) {
+        double max_dev = 0.0;
+        // loop all buses
+        for (Idx bus_number = 0; bus_number != this->n_bus_; ++bus_number) {
+            // Get maximum iteration for a bus
+            double const dev = max_val(cabs(updated_u_[bus_number] - u[bus_number]));
+            // Keep maximum deviation of all buses
+            max_dev = std::max(dev, max_dev);
+            // assign updated values
+            u[bus_number] = updated_u_[bus_number];
+        }
+        return max_dev;
+    }
+
+    void reset_lhs() {
```

Comment:
I have another idea to detect if the y_bus values has changed.

You don't need a separate `reset_lhs` function.

Reply:
The idea is:

1. You save the pointer to Y bus data into a class data member.
2. In your initialize step, you compare the pointer you retrieved from the current Y bus value and the internal cachec pointer.
3. If the pointers are the same, it means Y bus has not changed.
4. Otherwise it is changed.